diff -u -r include/ogg/os_types.h include/ogg/os_types.h
--- include/ogg/os_types.h
+++ include/ogg/os_types.h
@@ -138,6 +138,25 @@
    typedef unsigned int ogg_uint32_t;
    typedef long long int ogg_int64_t;
 
+#elif defined(ANDROID)
+
+#    include <sys/types.h>
+     typedef short ogg_int16_t;
+     typedef unsigned short ogg_uint16_t;
+     typedef int ogg_int32_t;
+     typedef unsigned int ogg_uint32_t;
+     typedef long long ogg_int64_t;
+     typedef unsigned long long ogg_uint64_t;
+
+#elif defined(__linux__)
+
+#    include <sys/types.h>
+     typedef short ogg_int16_t;
+     typedef unsigned short ogg_uint16_t;
+     typedef int ogg_int32_t;
+     typedef unsigned int ogg_uint32_t;
+     typedef long ogg_int64_t;
+
 #else
 
 #  include <ogg/config_types.h>
diff -u -r src/framing.c src/framing.c
--- src/framing.c
+++ src/framing.c
@@ -712,21 +712,21 @@
 
   /* yes, have a whole page all ready to go */
   {
-    unsigned char *page=oy->data+oy->returned;
-    long bytes;
+    unsigned char *page2=oy->data+oy->returned;
+    long bytes2;
 
     if(og){
-      og->header=page;
+      og->header=page2;
       og->header_len=oy->headerbytes;
-      og->body=page+oy->headerbytes;
+      og->body=page2+oy->headerbytes;
       og->body_len=oy->bodybytes;
     }
 
     oy->unsynced=0;
-    oy->returned+=(bytes=oy->headerbytes+oy->bodybytes);
+    oy->returned+=(bytes2=oy->headerbytes+oy->bodybytes);
     oy->headerbytes=0;
     oy->bodybytes=0;
-    return(bytes);
+    return(bytes2);
   }
 
  sync_fail:
